op-node: Skip L1 genesis hash check if block is not available from the L1

ajsutton opened the pull request:
**Description**

Skip L1 genesis hash check if block is not available from the L1.

With block and receipt data being pruned on a rolling basis from L1 nodes having a requirement that a very old, fixed block is present just to check a config that is often built into op-node itself seems unreasonable.

Without this, using Nethermind as the L1 with `--history-pruning=Rolling` (unclear if this is the default now) op-node fails to start with:
```
failed to init L1: failed to validate the L1 config: failed to get L1 genesis blockhash: failed to fetch header by num 17422590: not found
```

ajsutton committed:
op-node: Skip L1 genesis hash check if block is not available from the L1

sebastianst commented on op-node/rollup/types.go:
```diff
@@ -236,6 +238,10 @@ func (cfg *Config) CheckL1ChainID(ctx context.Context, client L1Client) error {
 	l1GenesisBlockRef, err := client.L1BlockRefByNumber(ctx, cfg.Genesis.L1.Number)
 	if err != nil {
+		if isNotFound(err) {
+			// Genesis block isn't available to check, so just accept it and hope for the best
```
Let's add a warn-log here.

ajsutton replied:
Done.

sebastianst commented on op-service/errutil/errors.go:
```diff
@@ -20,3 +23,11 @@ func TryAddRevertReason(err error) error {
+// IsEthereumNotFound determines if an error is likely to be ethereum.NotFound even if it has been serialized and
+// recreated through an RPC server (thus losing the specific typing). Since this depends on string matching, it may
+// return false positives.
+func IsEthereumNotFound(err error) bool {
```
We already have a function `op-service/eth.MaybeAsNotFoundErr` in file `errors.go`, which contains a very similar logic. We could combine both, probably moving `MaybeAsNotFoundErr` to the `errutil` package, then making it use `IsEthereumNotFound`. Note that the implementation of `MaybeAsNotFoundErr` contains even more cases:
```go
// MaybeAsNotFoundErr checks if the error is an ethereum.NotFound error
// or has an error string that heuristically indicates that it is this error.
// If so, it returns ethereum.NotFound, otherwise it returns the original error.
//
// Note that correct implementations of the execution layer API must return an empty
// result and no error if the block or header is not found. So using this translation
// hardens against wrong implementations of the execution layer API only.
func MaybeAsNotFoundErr(err error) error {
	if errors.Is(err, ethereum.NotFound) || err == nil {
		return err
	}
	if errStr := strings.ToLower(err.Error()); strings.Contains(errStr, "block not found") ||
		strings.Contains(errStr, "header not found") ||
		strings.Contains(errStr, "unknown block") {
		return errors.Join(err, ethereum.NotFound)
	}
	return err
}
```
So we may want to expand the implementation of `IsEthereumNotFound` to use this error match logic instead.

ajsutton replied:
`MaybeAsNotFoundErr` is very focussed on blocks and headers not being found and wouldn't actually pick up the not found error the supervisor returns when it can't find a super root.  Since these are actually different things to be looking for, I've switched the L1 genesis check to use `MaybeAsNotFoundErr` and put supervisor_client back to using its own `isNotFound`.

sebastianst reviewed (approved):
One suggestion to deduplicate with `op-service/eth.MaybeAsNotFoundErr`, otherwise LGTM.

ajsutton committed:
op-node: Switch L1 genesis check to using MaybeAsNotFoundErr

Switch supervisor_client back to it's own isNotFound. MaybeAsNotFoundErr is specific to identifying blocks and headers that are not found, the supervisor client is looking for super root not found. That's an important distinction so we should use separate code.

ajsutton committed:
Fix lint

ajsutton committed:
Update test
